libs/init/buildgen/src/mill/main/buildgen/BuildWriter.scala

```diff
@@ -0,0 +1,282 @@
+package mill.main.buildgen
+
+import mill.constants.CodeGenConstants.{nestedBuildFileNames, rootBuildFileNames, rootModuleAlias}
+import mill.internal.Util.backtickWrap
+import pprint.Util.literalize
+
+import java.io.PrintStream
+import scala.util.Using
+
+trait BuildWriter {
+
+  final def writeFiles(build: BuildRepr): Unit = {
+    import build.*
+    val root +: nested = packages.iterator.toSeq: @unchecked
+    Using.resource(PrintStream(os.write.outputStream(os.pwd / rootBuildFileNames.get(0)))):
+      writeRootPackage(_, root, baseTraits, depsObject)
+    nested.foreach: pkg =>
+      Using.resource(
+        PrintStream(os.write.outputStream(os.pwd / pkg.root.segments / nestedBuildFileNames.get(0)))
+      ):
+        writeNestedPackage(_, pkg)
+  }
+
+  protected def writeRootPackage(
+      ps: PrintStream,
+      pkg: Tree[ModuleRepr],
+      baseTraits: Seq[BaseTrait],
+      depsObject: Option[DepsObject]
+  ): Unit = {
+    ps.println(s"package $rootModuleAlias")
```

Comment:
Please use string concatenation rather than `println` for all this codegen

Reply:
Do you mean the entire type should be rendered as a string?

Reply:
yes